Optional HDF5 optimizations and H5timings for root serialization

QiaoK opened the pull request:
Contains the following features:

1. A bug fix for batch size. If the total number of objects is not divisible by the batch size, some objects are not written. This issue is fixed.
2. Arguments for controling the batch size and HDF5 method
3. Two additional HDF5 strategies: aggregation and H5O implementation.
4. H5Timing support for time details (disabled by default)
5. A typo preventing compiling in the existing master branch
6. Incorporated Dr. Jones comment for merging. Minimum changes to existing code. Most variables are moved to extern multidataset_plugin library.

Dr15Jones commented:
@QiaoK it appears that you are much more comfortable programming in the stye of C instead of C++. I encourage you to take working on this project as an opportunity to increase your comfort in using modern C++ best practices. @knoepfel @sabasehrish and I are willing to help in that regard if you are interested.

Dr15Jones commented on HDFOutputer.cc:
```diff
@@ -115,20 +137,59 @@ HDFOutputer::output(EventIdentifier const& iEventID,
 
   ++batch_;
-  if (batch_ == maxBatchSize_) {
+  if (total_n_events > 0) {
```
As describe elsewhere in the review, this isn't a general enough solution as the total number of events is not actually known. Just the upper bound.

Dr15Jones commented on multidataset_plugin.cc:
```diff
@@ -0,0 +1,544 @@
+    dataset_size = 0;
+    dataset_size_limit = 0;
+
+    char *p = getenv("HEP_MAX_BATCH_SIZE");
```
Please do not use environment variables to tune. These should come from the command line arguments passed in to the program. An easy modification is to pass the values via parameters of this function.

Dr15Jones reviewed (changes requested):
I'm available for a video conference to discuss possible ways to make changes.

Dr15Jones commented on multidataset_plugin.h:
```diff
@@ -0,0 +1,65 @@
+#define ENABLE_MULTIDATASET 0
+#define MULTIDATASET_DEFINE 1
+
+static int max_batch_size_g = 2;
+static int hdf5_method_g = -1;
```
Instead of having these be global variables, how about moving them to an struct? It could be returned from the call to `init_multidataset()` and then passed to all functions needing the variables. That would guarantee that `init_` is actually called before any of the functions using its side-effect. It also avoids the use of global data which makes reasoning about thread-safety much harder.

Dr15Jones commented on multidataset_plugin.h:
```diff
@@ -0,0 +1,65 @@
+
+static int max_batch_size_g = 2;
+static int hdf5_method_g = -1;
+static int total_n_events_g = -1;
```
Can this variable go away now?

Dr15Jones commented on multidataset_plugin.cc:
```diff
@@ -0,0 +1,544 @@
+#include "multidataset_plugin.h"
+
+
+static multidataset_array *multi_datasets;
+/*
+static hid_t *dataset_recycle;
+static hid_t *memspace_recycle;
+static hid_t *dataspace_recycle;
+*/
+static int dataset_size;
+static int dataset_size_limit;
+
+/*
+static int dataset_recycle_size;
+static int dataset_recycle_size_limit;
+static int dataspace_recycle_size;
+static int dataspace_recycle_size_limit;
+static int memspace_recycle_size;
+static int memspace_recycle_size_limit;
+*/
+hsize_t total_data_size;
```
How about moving all these global variables to a struct returned from the `init_multidataset()` call?

Dr15Jones commented:
@QiaoK Let's see if we can move things forward. How much of the code is from existing work you did for other projects? I'd want to avoid having to make any changes to that code in particular. Only the parts that are needed to work with this repository would I request further changes.

QiaoK commented:
I can have another commit addressing these comments. This will take about a
day. These days I am heavily catching another project. Please pardon me for
the delay.

On Tue, Mar 22, 2022 at 11:28 AM Chris Jones ***@***.***>
wrote:

> @QiaoK <https://github.com/QiaoK> Let's see if we can move things
> forward. How much of the code is from existing work you did for other
> projects? I'd want to avoid having to make any changes to that code in
> particular. Only the parts that are needed to work with this repository
> would I request further changes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/hep-cce2/root_serialization/pull/27#issuecomment-1075481933>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABLXTOSV7GE5J7HJLXSIJPTVBIGNFANCNFSM5O5SRYOA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Dr15Jones commented:
@QiaoK no problem with any delays, we are all very busy now-a-days :). I also wanted to reduce some of the change burden I had requested which is why I asked about code included here that was part of another project. I don't want to have to have you change code you are sharing across projects if at all possible.

Dr15Jones commented:
@QiaoK I'm afraid I've made several changes to the master since you started work on this pull request. It means you will need to rebase your work onto the present master. When doing so, I've learned from experience it is best to first make a local copy of your branch before attempting to do the rebase on the branch itself. That gives you an easy way to recover if something goes wrong.

QiaoK commented:
Hello Chirs,

I have implemented everything using C++ vector and map instances. This should avoid the use of manual incrementation of arrays. For register event data, it is true that map could improve the performance slightly.

I have resolved the conflicts. There is a small issue I need to ask for you. Is there a direct way to get the total number of events specified by users in the main function? I am not sure where this value is parsed from user command line. Previously I took the value from command line arguments, but it seems to be that piece of code is removed. I need that information for HDF outputer in order to make  sure the last batch can finish.

QiaoK commented:
No worry. I have merged my patch to your fix. writeBatch() function should perform all batch I/O calls.

Currently, the only part I need to touch the master branch is 4 lines of code in the main() function of threaded_io_test.

In addition, the writeBatch() function chooses I/O method according to user's will. By default, it is exactly the same as the master branch. H5Timing.h is disabled by default, so there is no need to worry about that.

I think now the merge is good to go. If you want me to start another pull request to show the code differences more clearly, I can also do that.

Dr15Jones commented on CMakeLists.txt:
```diff
@@ -52,6 +54,8 @@ target_link_libraries(batchevents_classes_dict PUBLIC ROOT::RIO ROOT::Net)
 
 add_executable(threaded_io_test
   DeserializeStrategy.cc
+  multidataset_plugin.cc
+  H5Timing.cc
```
please move these down to the section at the bottom which enables HDF5

Dr15Jones commented on CMakeLists.txt:
```diff
@@ -96,6 +100,8 @@ target_link_libraries(threaded_io_test
                               Threads::Threads
                               configKeys
                               sequence_classes_dict
+                              hdf5_hl
+                              hdf5
```
Please move these down to the enable HDF5 section

Dr15Jones commented on threaded_io_test.cc:
```diff
@@ -51,6 +57,11 @@ int main(int argc, char* argv[]) {
     return 1;
   }
 
+#ifdef H5_TIMING_ENABLE
```
Again, please remove all changes you made to this function and put the needed calls into HDFOutputer as that is the code which requires this setup and teardown. Especially since we need to be able to build threaded_io_test.cc even when we have no HDF libraries.

QiaoK commented:
Could you please let me know the entry function of HDFouputer (the very beginning where it is initialized)?

Dr15Jones commented:
> Could you please let me know the entry function of HDFouputer

That would be the constructor, HDFOutputer::HDFOutputer. The body of the constructor is on line 64 on your version of the file.

QiaoK commented:
> > Could you please let me know the entry function of HDFouputer
> 
> That would be the constructor, HDFOutputer::HDFOutputer. The body of the constructor is on line 64 on your version of the file.

Thanks. Changes are made.

Dr15Jones commented on multidataset_plugin.h:
```diff
@@ -0,0 +1,47 @@
+    hid_t did;
+    hid_t mtype;      /* memory datatype ID */
+    std::vector<char*> *temp_mem;
+} multidataset_array;
```
It would be better of all of these vectors were just held by value instead of by pointer. The compiler would then handle all the memory management.

Dr15Jones commented on threaded_io_test.cc:
```diff
@@ -18,6 +18,12 @@
 #include "tbb/task_group.h"
 #include "tbb/global_control.h"
 #include "tbb/task_arena.h"
+#include "H5Timing.h"
+#include "multidataset_plugin.h"
+
+extern int max_batch_size;
+extern int hdf_method;
+extern int total_n_events;
```
It looks like all of this can now go away.

Dr15Jones commented:
@QiaoK it seems like you are unfamiliar with a number of the C++ concepts. As we'd like to get your changes in for others to try, I have a proposal for two ways to proceed.
1. I continue giving review comments and you make the suggested changes you agree with, or

2. we merge your changes and I then rework them to follow more of the standard C++ idioms.

Which would you prefer?

QiaoK commented:
> @QiaoK it seems like you are unfamiliar with a number of the C++ concepts. As we'd like to get your changes in for others to try, I have a proposal for two ways to proceed.
> 
> 1. I continue giving review comments and you make the suggested changes you agree with, or
> 2. we merge your changes and I then rework them to follow more of the standard C++ idioms.
> 
> Which would you prefer?

Hello Chris,

Thank you for you comments. I think we can have a checkpoint now as this implementation is stable. Then I can have a separate pull request to work with you later with the C++ changes. I believe I can learn a lot from you. These days I have been catching a another deadline, but I can start this in the coming week I think.

Dr15Jones commented:
@QiaoK let's merge this pull request after you have removed the unnecessary `#include` and `extern` directives in threaded_io_test.cc.

QiaoK commented:
> @QiaoK let's merge this pull request after you have removed the unnecessary `#include` and `extern` directives in threaded_io_test.cc.

These have been removed accordingly.
